src/add_two_numbers.rs

```diff
@@ -0,0 +1,87 @@
+/*
+ * https://leetcode.com/problems/add-two-numbers/
+ *
+ * First: Definitions directly from leetcode.
+ */
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct ListNode {
+    pub val: u32,
+    pub next: Option<Box<ListNode>>,
+}
+
+impl ListNode {
+    #[inline]
+    fn new(val: u32) -> Self {
+        ListNode { next: None, val }
+    }
+}
+/*
+ * Solution:
+ */
+pub fn add_two_numbers(
+    l1: Option<Box<ListNode>>,
+    l2: Option<Box<ListNode>>,
+) -> Option<Box<ListNode>> {
+    full_adder(l1, l2, 0)
+}
+
+fn full_adder(
+    left: Option<Box<ListNode>>,
+    right: Option<Box<ListNode>>,
+    carry: u32,
+) -> Option<Box<ListNode>> {
+    if left.is_none() && right.is_none() && carry == 0 {
+        return None;
+    }
+    let l = left.as_ref().map_or(0, |b| b.val);
+    let r = right.as_ref().map_or(0, |b| b.val);
+    let result = l + r + carry;
+
+    let mut to_return = ListNode::new(result % 10);
+    to_return.next = full_adder(
+        left.and_then(|b| b.next),
```

Comment:
_Small_ idiom tweak: since you have a method that itself returns an option, you can just use `map`. If you want to, you could put, say, an associated (Rust-ese for static) method on your `impl` and just pass it:

```rust
fn next(ln: ListNode) -> Option<Box<ListNode> {
  ln.next
}

// later

left.map(ListNode::next)
```

Re-jiggering the lambda is entirely up to you; `map` is more idiomatic with functions or attributes that already return an `Option`.

Reply:
(It doesn't even really have to be on the impl, since it doesn't call any other private methods -- just a pattern I like for grouping methods under the type they act on. You can absolutely do the same thing with a `mod`, or leave 'em free floating.)